drivers/regulator/regulator_common.c

```diff
@@ -27,12 +29,35 @@ int regulator_common_init(const struct device *dev, bool is_enabled)
 		}
 	}
 
-	/* regulator voltage needs to be within allowed range before enabling */
+	/* If we have valid range values, we try to match them before enabling */
 	if ((config->min_uv > INT32_MIN) || (config->max_uv < INT32_MAX)) {
-		ret = regulator_set_voltage(dev, config->min_uv,
-					    config->max_uv);
-		if ((ret < 0) && (ret != -ENOSYS)) {
-			return ret;
+
+		min_uv = config->min_uv;
+		max_uv = config->max_uv;
+		apply_uv = true;
+
+		ret = regulator_get_voltage(dev, &current_uv);
+		if (ret < 0) {
+			if (ret != -ENOSYS) {
+				return ret;
+			}
+		} else {
+
+			/* Snap to closest interval value if out of range */
+			if (current_uv < min_uv) {
+				max_uv = min_uv;
+			} else if (current_uv > max_uv) {
+				min_uv = max_uv;
+			} else {
+				apply_uv = false;
+			}
+		}
+
+		if (apply_uv) {
+			ret = regulator_set_voltage(dev, min_uv, max_uv);
+			if ((ret < 0) && (ret != -ENOSYS)) {
+				return ret;
+			}
 		}
```

Comment:
this code has some issues, min/max range doesn't necessarily mean the voltage can be exactly synthesized, e.g. you could specify a maximum of 5V in regulator-max-voltage, but regulator may be able to generate 4.95, 5.05, etc. In the first _snap_ branch, you'll end up with max_uv = min_uv, meaning set_voltage will fail in the condition described previously. I think the code below would achieve what you want (not tested):

```suggestion
		int32_t current_uv;

		ret = regulator_get_voltage(dev, &current_uv);
		if ((ret < 0) && (ret != -ENOSYS)) {
			return ret;
		}

		/* apply voltage if current one is out of range */
		if ((current_uv < config->min_uv) || (current_uv > config->max_uv)) {
			ret = regulator_set_voltage(dev, config->min_uv, config->max_uv);
			if ((ret < 0) && (ret != -ENOSYS)) {
				return ret;
			}
		}
```

Reply:
> this code has some issues, min/max range doesn't necessarily mean the voltage can be exactly synthesized, e.g. you could specify a maximum of 5V in regulator-max-voltage, but regulator may be able to generate 4.95, 5.05, etc. In the first _snap_ branch, you'll end up with max_uv = min_uv, meaning set_voltage will fail in the condition described previously. I think the code below would achieve what you want (not tested):

While this may happen, I see this more of a devicetree configuration issue. Given that the HW configuration is closely tied with the dts configuration, the min/max constraints should be set accordingly in the dts, given the (known) device you are going to use. If this is not the case, you will end up with a `!device_is_ready()` condition, as the regulator will fail `init()` if the voltage could not be set exactly at the interval edges.

Seems like linux does it the same way:

```
if (current_uV < rdev->constraints->min_uV) {
	target_min = rdev->constraints->min_uV;
	target_max = rdev->constraints->min_uV;
}

if (current_uV > rdev->constraints->max_uV) {
	target_min = rdev->constraints->max_uV;
	target_max = rdev->constraints->max_uV;
}
```

The solution you proposed fails in case the regulator returns `-ENOSYS` as `current_uv` will contain an undefined value.
Also, the idea would be to snap the voltage at the closest range value. Setting it with `config->min_uv` and `config->max_uv` will always end up at the lowest value.

Reply:
I'm inclined to agree with @gmarull here, that we should support the case where the user's devicetree specifies a voltage range, and the regulator makes the best effort possible to align with that selection.

For example, what about some type of device that requires several power rails? If the regulator attempts to snap to the closest supported value, then the devicetree describing that device can be written as an overlay, like this:

```
my_dev: device@0 {
	compatible = "vnd,my-dev";
	reg = <0>;
};

&rail0 {
	regulator_max_microvolt = <5000000>;
	regulator_min_microvolt = <4500000>;
};

&rail1 {
	regulator_max_microvolt = <3300000>;
	regulator_min_microvolt = <3000000>;
};
```

That overlay can then be used across multiple boards, provided they all assign nodelabels for `rail0` and `rail1`

Reply:
> The solution you proposed fails in case the regulator returns `-ENOSYS` as `current_uv` will contain an undefined value. 

Adjusted to handle -ENOSYS properly. Sorry about that.

> Also, the idea would be to snap the voltage at the closest range value. Setting it with `config->min_uv` and `config->max_uv` will always end up at the lowest value.

That's a valid concern. We need to find what is best here, whether to treat `regulator-min/max-microvolt` as synthesizable voltages or not. By looking at the API (set_voltage), we provide a range there to make code more portable, ie, if my circuit tolerates 3-3.3V, I don't really care what the regulator sets as soon as it's within that range. But in Devicetree, we may force users to put valid voltages in there. The downside, as @danieldegrasse says is that we can't then have fully portable overlays. But it's true that _most_ regulators are designed to generate _nice_ voltage levels, ie, you probably won't find any regulator that can't generate 5.000V. It's more about the step resolution. So in the end, overlays may end up being quite portable despite this limitation.

I'm now inclined to agree with the proposal, ie, align with Linux. I'd simplify the code a bit though:

```suggestion
		ret = regulator_get_voltage(dev, &current_uv);
		if (ret < 0) {
			return ret;
		}

		/* snap to closest interval value if out of range */
		if (current_uv < config->min_uv) {
			ret = regulator_set_voltage(dev, config->min_uv, config->min_uv);
			if (ret < 0) {
				return ret;
			}
		} else if (current_uv > config->max_uv) {
			ret = regulator_set_voltage(dev, config->max_uv, config->max_uv);
			if (ret < 0) {
				return ret;
			}	
		}
```

Reply:
Sorry for the late response

@gmarull Agree. I've added the changes.

@danieldegrasse
While having a generic overlay for an application targeting multiple boards might be a use-case, as @gmarull pointed, most regulators, if adjustable, typically support most common voltage snap points.

I don't see any obvious non-overkill way to address this issue as a search would need to happen to find the actual edge.